### PR TITLE
General Layout Support

### PR DESCRIPTION
**Pull request by yzhliu**
This PR allows NNVM to do
* Replace an operator via `AlterOpLayout`. For x86 it is more efficient to calculate convolution in `NCHW16c` layout. Here is an example of how it is used: https://github.com/yzhliu/topi-intel/blob/master/e2e_general_pack/schedule_pack/avx512_conv_fwd.py#L128-L155
Note that kernel (pre-)packing is supported as well: https://github.com/dmlc/nnvm/issues/303

* Infer and correct layout automatically
  - Given the input data layout, or the layout of operators in the network (e.g., convolution, pooling...) `InferCorrectLayout` pass can infer the layout for each operator (both inputs and outputs).
  - If the required input layout of an operator is different from what it receives, a `__layout_transform__` operator will be inserted.
  - Each operator registers a function `FInferLayout`. Once a model is imported,  we do a `InferCorrectLayout` pass and store the *original layouts* for each operator. After `AlterOpLayout`, the `InferCorrectLayout` runs again. This time each operator will see the *original layouts* it inferred before, which can be used to decide whether to keep the original one. For example, `softmax` can still generate correct result after the input layout changes, while `flatten` cannot. So `flatten` claims it needs the `original layout` and triggers a layout transform, which make the network produce correct result.
  - With such approach,
     * optimized layout (e.g., NCHW16c) can flow through the network as far as possible. No pack/unpack overhead for each (e.g., convolution) operator.
     * now model layout can be transparent to users. Even though a convolution neural network is trained with `NHWC` layout, users can still pass `NCHW` input as long as the input layout is specified, a layout transform happens automatically.

Moreover the convolution kernel layout becomes much clearer: https://github.com/dmlc/nnvm/pull/372 . Now we have kernel layout `OIHW` and `HWIO` for `NCHW` and `NHWC` respectively.

Will shoot a pull request for corresponding TVM changes.

@yidawang @kevinthesun

### CODE REVIEW

File: python/nnvm/compiler/build_module.py [hunk 1]
```diff
@@ -204,8 +218,8 @@ def build(graph, target=None, shape=None, dtype="float32", params=None, target_h
         By default, llvm is used if it is enabled,
         otherwise a stackvm intepreter is used.
 
-    initialize : bool, optional
-        Whether to initialize variables in global dict _all_var_init.
+    layout : dict of str to str or str
```
Comment (kevinthesun) [inline python/nnvm/compiler/build_module.py]: Does user need to specify opt_level=3 to enable layout transform? Also how does user specify "internal" layout, if the target layout is not NCHWXc?

File: src/top/elemwise_op_common.h [hunk 2]
```diff
@@ -100,12 +103,176 @@ inline bool ElementWiseReduceType(const NodeAttrs& attrs,
...
+          << name << ": " << "expected " << *target
+          << ", got " << vec->at(i);
+      }
+    }
+  };
+
+  Layout in, last_in, out;
+  deduce(&in, in_layouts, in_size, "input");
+  deduce(&last_in, last_in_layouts, in_size, "input (last infer pass)");
+  deduce(&out, out_layouts, out_size, "output");
```
Comment (kevinthesun) [inline src/top/elemwise_op_common.h]: Assume an elemwise_add op has two inputs with layout "NCHW16c" and "NCHW4c", will they be transformed back to "NCHW", or one of them will be transformed to another?

Comment (yzhliu) [reply src/top/elemwise_op_common.h]: right transform to left. it's defined in `ElemwiseBinaryKeepLeftLayout`

File: python/nnvm/frontend/mxnet.py [hunk 3]
```diff
@@ -237,7 +247,6 @@ def _upsampling(inputs, attrs):
     'min_axis'      : _rename('min'),
     'reshape'       : _reshape,
     'sum_axis'      : _rename('sum'),
-    'UpSampling'    : _upsampling
```
[the remark this answers is not in the document]
Comment (yzhliu) [reply python/nnvm/frontend/mxnet.py]: oh I was removing SSD stuff but by mistake remove this...

File: include/nnvm/compiler/packed_func_ext.h [hunk 4]
```diff
@@ -52,6 +54,12 @@ template<>
 struct extension_class_info<nnvm::compiler::AttrDict> {
   static const int code = 18;
 };
+
+template<>
+struct extension_class_info<nnvm::compiler::SymbolArray> {
+  static const int code = 19;
```
Comment (tqchen) [inline include/nnvm/compiler/packed_func_ext.h]: cannot use 19, because it is taken by mxnet tvm bridge https://github.com/apache/incubator-mxnet/blob/master/include/mxnet/tensor_blob.h#L49

File: include/nnvm/compiler/op_attr_types.h [hunk 5]
```diff
@@ -73,19 +74,17 @@ using FTVMSchedule = std::function<
...
+ * \param attrs The attribute of the original node.
+ * \param inputs The input symbols of the original node.
+ * \param tinfos The inferred shape and dtype of the inputs.
  */
-using FTVMLayoutRequest = std::function<bool (const NodeAttrs& attrs,
-                                              std::vector<TLayoutInfo> *ilayouts,
-                                              std::vector<TLayoutInfo> *olayouts)>;
+using FTVMAlterOpLayout = std::function<
+  Symbol(const NodeAttrs& attrs,
+         const SymbolArray& inputs,
```
Comment (tqchen) [inline include/nnvm/compiler/op_attr_types.h]: do we really need the input symbols, or is tinfo enough for now?

File: include/nnvm/compiler/packed_func_ext.h (hunk 4 above)
Comment (tqchen) [reply include/nnvm/compiler/packed_func_ext.h]: It seems to be quite dangerous to expose a vector of pointers here.

File: include/nnvm/compiler/op_attr_types.h [hunk 6]
```diff
@@ -73,19 +74,17 @@ using FTVMSchedule = std::function<
...
+ *  it is invoked in AlterOpLayout pass.
+ * \param attrs The attribute of the original node.
+ * \param inputs The input symbols of the original node.
+ * \param tinfos The inferred shape and dtype of the inputs.
  */
-using FTVMLayoutRequest = std::function<bool (const NodeAttrs& attrs,
-                                              std::vector<TLayoutInfo> *ilayouts,
-                                              std::vector<TLayoutInfo> *olayouts)>;
+using FTVMAlterOpLayout = std::function<
+  Symbol(const NodeAttrs& attrs,
```
Comment (tqchen) [inline include/nnvm/compiler/op_attr_types.h]: Can we just return the result output layout as vector of string?

File: include/nnvm/compiler/op_attr_types.h (hunk 5 above)
Comment (tqchen) [reply include/nnvm/compiler/op_attr_types.h]: ignore this comment, I now get what is going on

File: src/pass/infer_correct_layout.cc [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+  return n;
+}
+
+using LayoutAttrDict = std::unordered_map<const Node*, std::vector<Layout> >;
+
+/*!
+ * \brief A simple layout infer pass that will
+ *        insert layout transform nodes automatically.
+ */
+nnvm::Graph InferCorrectLayout(nnvm::Graph src) {
```
Comment (tqchen) [inline src/pass/infer_correct_layout.cc]: Maybe we can just call it CorrectLayout

File: python/nnvm/top/symbol_array.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
+# pylint: disable=invalid-name
+"""Symbol array object used by weight pre-pack"""
+import tvm
+
+_sym_arr_get = tvm.get_global_func("nnvm.compiler._symbol_array_get")
+_sym_arr_size = tvm.get_global_func("nnvm.compiler._symbol_array_size")
+
+class SymbolArray(object):
```
Comment (tqchen) [inline python/nnvm/top/symbol_array.py]: We can just use symbol for this, because we can group multiple nodes into a single Symbol

File: include/nnvm/compiler/op_attr_types.h (hunk 6 above)
Comment (tqchen) [reply include/nnvm/compiler/op_attr_types.h]: ignore this

File: src/pass/infer_correct_layout.cc (hunk 7 above)
Comment (yzhliu) [reply src/pass/infer_correct_layout.cc]: ok. will change.

File: include/nnvm/layout.h [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+#include <sstream>
+#include <vector>
+#include <utility>
+#include <algorithm>
+
+namespace nnvm {
+
+class Layout {
+ public:
+  using LayoutDim = char;
```
Comment (yzhliu) [inline include/nnvm/layout.h]: @tqchen 
Do you want me to change the name 'dim' to 'axis' as you suggested in https://discuss.tvmlang.org/t/datalayout-structure/80/3 ? 

And to my understanding, we don't have to wait for that DataLayout implemented in TVM. We can merge this PR first (so that we can move forward to having significant improvement - at least 50% speedup - for CNN) and switch to that in TVM later, right?

Comment (tqchen) [reply include/nnvm/layout.h]: yes, we can merge this in first, please fix the problems I mentioned

**Comment by tqchen** [comment]
@merrymercy Can you also do a review when you have time?

**Comment by yzhliu** [comment]
@tqchen The segment fault in CI looks weird. I cannot reproduce. And it happened in PR #435 as well: http://mode-gpu.cs.washington.edu:8080/blue/organizations/jenkins/dmlc%2Fnnvm/detail/PR-435/13/pipeline

File: src/pass/correct_layout.cc [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    for (uint32_t i = 0; i < inode.inputs.size(); ++i) {
+      const auto& e = inode.inputs[i];
+      const nnvm::NodePtr& in = mirror_vec[e.node_id];
+      new_node->inputs[i] = nnvm::NodeEntry{in, e.index, e.version};
+
+      // insert layout_transform if necessary
+      const Layout& produce = produce_ilayouts[i];
+      const Layout& request = request_ilayouts[i];
+      if (produce != request && produce.defined()) {
+        LOG(INFO) << "Insert layout transformer for " << i << "-th input of "
```
Comment (tqchen) [inline src/pass/correct_layout.cc]: This seems to produce a lot of messages, maybe avoid print this, since this is expected behavior of the pass

**Comment by tqchen** [comment]
Thanks for improving the code through the review! this is now merged
